src/core_plugins/elasticsearch/lib/patch_kibana_index.js

```diff
@@ -0,0 +1,101 @@
+import {
+  getTypes,
+  getRootType,
+  getRootProperties
+} from '../../../server/mappings';
+
+/**
+ *  Checks that the root type in the kibana index has all of the
+ *  root properties specified by the kibanaIndexMappings.
+ *
+ *  @param  {Object} options
+ *  @property {Function} options.log
+ *  @property {string} options.indexName
+ *  @property {Function} options.callCluster
+ *  @property {EsMappingsDsl} options.kibanaIndexMappingsDsl
+ *  @return {Promise<undefined>}
+ */
+export async function patchKibanaIndex(options) {
+  const {
+    log,
+    indexName,
+    callCluster,
+    kibanaIndexMappingsDsl
+  } = options;
+
+  const rootEsType = getRootType(kibanaIndexMappingsDsl);
+  const currentMappingsDsl = await getCurrentMappings(callCluster, indexName, rootEsType);
+  const missingProperties = await getMissingRootProperties(currentMappingsDsl, kibanaIndexMappingsDsl);
+
+  const missingPropertyNames = Object.keys(missingProperties);
+  if (!missingPropertyNames.length) {
+    // all expected properties are in current mapping
+    return;
+  }
+
+  // log about new properties
+  log(['info', 'elasticsearch'], {
+    tmpl: `Adding mappings to kibana index for SavedObject types "<%= names.join('", "') %>"`,
+    names: missingPropertyNames
+  });
+
+  // add the new properties to the index mapping
+  await callCluster('indices.putMapping', {
+    index: indexName,
+    type: rootEsType,
+    body: {
+      properties: missingProperties
+    }
+  });
+}
+
+/**
+ *  Get the mappings dsl for the current Kibana index
+ *  @param  {Function} callCluster
+ *  @param  {string} indexName
+ *  @param  {string} rootEsType
+ *  @return {EsMappingsDsl}
+ */
+async function getCurrentMappings(callCluster, indexName, rootEsType) {
+  const index = await callCluster('indices.get', {
+    index: indexName,
+    feature: '_mappings'
+  });
+
+  // could be different if aliases were resolved by `indices.get`
+  const resolvedName = Object.keys(index)[0];
+  const currentMappingsDsl = index[resolvedName].mappings;
+  const currentTypes = getTypes(currentMappingsDsl);
+
+  const isV5Index = currentTypes.length > 1 || currentTypes[0] !== rootEsType;
+  if (isV5Index) {
+    throw new Error(
+      'Your Kibana index is out of date, reset it or use the X-Pack upgrade assistant.'
```

Comment:
The idea with `reset` is for people who are alright just deleting their Kibana index, primarily devs who will see this as soon as they pull master... I'm not dead-set on shipping with the verbiage though